pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/KeyValueSchemaCompatibilityCheck.java

```diff
@@ -0,0 +1,138 @@
+ */
+package org.apache.pulsar.broker.service.schema;
+
+import com.google.gson.Gson;
+import org.apache.pulsar.common.schema.KeyValue;
+import org.apache.pulsar.common.schema.SchemaData;
+import org.apache.pulsar.common.schema.SchemaType;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * {@link KeyValueSchemaCompatibilityCheck} for {@link SchemaType#KEY_VALUE}.
+ */
+public class KeyValueSchemaCompatibilityCheck implements SchemaCompatibilityCheck {
+
+    private final Map<SchemaType, SchemaCompatibilityCheck> checkers;
+
+    public KeyValueSchemaCompatibilityCheck(Map<SchemaType, SchemaCompatibilityCheck> checkers) {
+        this.checkers = checkers;
+    }
+
+    private KeyValue<byte[], byte[]> splitKeyValueSchemaData(byte[] bytes) {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+        int keyLength = byteBuffer.getInt();
+        byte[] keySchema = new byte[keyLength];
+        byteBuffer.get(keySchema);
+
+        int valueLength = byteBuffer.getInt();
+        byte[] valueSchema = new byte[valueLength];
+        byteBuffer.get(valueSchema);
+        return new KeyValue<>(keySchema, valueSchema);
+    }
+
+    private SchemaType fetchSchemaType(Map<String, String> properties, String key) {
+        if (properties.get(key) != null) {
+            return SchemaType.valueOf(properties.get(key));
+        }
+        return SchemaType.BYTES;
+    }
+
+    private SchemaData fetchSchemaData(
+            byte[] keyValue, SchemaType schemaType, Gson schemaGson, Map<String, String> properties, String key) {
+        if (properties.get(key) != null) {
+            return SchemaData.builder().data(keyValue)
+                    .type(schemaType)
+                    .props(schemaGson.fromJson(properties.get(key), Map.class)).build();
+        }
+        return SchemaData.builder().data(keyValue)
+                .type(schemaType)
+                .props(Collections.emptyMap()).build();
+    }
+
+    private KeyValue<SchemaData, SchemaData> parseSchemaData(SchemaData schemaData) {
+        KeyValue<byte[], byte[]> keyValue = this.splitKeyValueSchemaData(schemaData.getData());
+        Map<String, String> properties = schemaData.getProps();
+        SchemaType keyType = fetchSchemaType(properties, "key.schema.type");
+        SchemaType valueType = fetchSchemaType(properties, "value.schema.type");
+        Gson schemaGson = new Gson();
+        SchemaData keySchemaData = fetchSchemaData(
+                keyValue.getKey(), keyType, schemaGson, properties, "key.schema.properties");
+        SchemaData valueSchemaData = fetchSchemaData(
+                keyValue.getValue(), valueType, schemaGson, properties, "value.schema.properties");
+        return new KeyValue<>(keySchemaData, valueSchemaData);
+    }
+
+    @Override
+    public SchemaType getSchemaType() {
+        return SchemaType.KEY_VALUE;
+    }
+
+    @Override
+    public boolean isWellFormed(SchemaData to) {
```

Comment:
I don't think the logic is correct here.

You need to get key schema check and use that to check key schema, and get value schema check to check value schema. You are using same`valueCheck` for checking both key and value schema.